### PR TITLE
Embedding + annotation pipeline at scale

### PR DESCRIPTION
**Pull request by ajude2s**
### Summary
This PR introduces two complementary pipelines for embedding generation and annotation, built on top of the **DataTrove** framework. Together, they enable a fully modular, efficient, and scalable workflow for processing large text datasets — from creating embeddings to performing embedding-based annotation and evaluation.

---

### 🔹 Embedding Pipeline
A configurable embedding pipeline that:

- **Reads** textual data from `.jsonl` files using `JsonlReader`.  
- **Embeds** documents using a configurable embedding model via `JQLEmbedder` to compute dense representations.  
- **Writes** results in structured **HDF5 (.h5)** format, optimized for large-scale processing and training.  
- **Executes** via `LocalPipelineExecutor` or `SlurmPipelineExecutor`, supporting batch and parallel processing for performance and scalability.  

**Testing:**  
Includes integration and unit tests to validate embedding generation, I/O consistency, and overall pipeline behavior.  

**Motivation:**  
Simplifies and scales the embedding process for large datasets, ensuring fast, consistent, and reproducible access to embeddings for downstream ML workflows.

---

### 🔹 Annotation Pipeline
A modular annotation pipeline that operates directly on precomputed embeddings:

- **JQLEmbeddingReader** – Loads embeddings and labels from `.h5` files.  
- **JQLHead** – Applies regression heads to embeddings for annotation.  
- **stats_adapter** – Extracts and flattens metadata for output.  
- **run_annotation_pipeline** – Configures and executes the full pipeline using `LocalPipelineExecutor` or `SlurmPipelineExecutor`.  

**Testing:**  
Comprehensive unit and integration tests validate component functionality and verify successful end-to-end pipeline execution, including JSONL output generation.  

**Motivation:**  
Establishes the foundation for scalable, embedding-based annotation and evaluation workflows, enabling efficient filtering and analysis of large datasets.

### CODE REVIEW

File: configs/annotation/lorem_ipsum_embedding.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
+dataset_name: training
+
+params:
+  # File selection
+  glob_pattern: "**/*.jsonl"
+  input_dir: /home/abbas-khan/processed_data_natural/${dataset_name}_set
+  csv_hashmap_path: /home/abbas-khan/ml_filter/data/output/hashes/${dataset_name}.csv
```
Comment (AbasKhan) [inline configs/annotation/lorem_ipsum_embedding.yaml]: We dont need this anymore right ?

File: src/ml_filter/annotation/datatrove_jql_annotator.py [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+    name = "🔢 - JQL-EMBEDDER"
+    type = "🔢 - JQL-EMBEDDER"
+
+    def __init__(
+        self,
+        embedder_model_id: str,
+        batch_size: int = 1_000,
+        device_overwrite: Optional[str] = None,
+        stats_writer: DiskWriter = None,
+        max_length: int = 8192,
```
Comment (AbasKhan) [inline src/ml_filter/annotation/datatrove_jql_annotator.py]: I would rather not give these defaults, these values should always be specified in the configs

File: src/ml_filter/annotation/datatrove_jql_annotator.py [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+            emb_ds = group["embeddings"]
+            id_ds = group["document_id"]
+            labels_ds = group.get("labels") if "labels" in group else None
+
+        # Resize for new batch once and append
+        old_size = emb_ds.shape[0]
+        new_size = old_size + embeddings.shape[0]
+        emb_ds.resize(new_size, axis=0)
+        id_ds.resize(new_size, axis=0)
+        emb_ds[old_size:new_size, :] = embeddings
```
Comment (AbasKhan) [inline src/ml_filter/annotation/datatrove_jql_annotator.py]: Maybe changing the names of these variable to something more appropriate ? . and Maybe some comment explaining this would be nice

File: src/ml_filter/annotation/embedder.py [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+            # For right padding, find the last valid token position for each sequence
+            # attention_mask.sum(dim=1) gives the length of each sequence
+            # Subtract 1 to get the index of the last valid token (0-indexed)
+            sequence_lengths = attention_mask.sum(dim=1) - 1
+            batch_size = last_hidden_states.shape[0]
+
+            # Select the last valid token for each sequence
+            return last_hidden_states[torch.arange(batch_size, device=last_hidden_states.device), sequence_lengths]
+
+    def get_detailed_instruct(self, task_description: str, query: str) -> str:
```
Comment (AbasKhan) [inline src/ml_filter/annotation/embedder.py]: This isnt used anywhere now right ?

File: src/ml_filter/annotation/regression_head.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
+from torch import nn
+import lightning as L
+
+class RegressionHead(L.LightningModule):
```
Comment (AbasKhan) [inline src/ml_filter/annotation/regression_head.py]: DO we need this ? , embedding training branch has an implementation aswell
`https://github.com/Modalities/ml_filter/blob/regression_exp_abbas_embed/src/ml_filter/models/annotator_model_head.py`

File: src/ml_filter/annotation/datatrove_jql_annotator.py (hunk 2 above)
Comment (ajude2s) [reply src/ml_filter/annotation/datatrove_jql_annotator.py]: Removed the defaults in 5eef40ed6d1f6622ca7e2cffa56bcd95878164d3

File: src/ml_filter/annotation/datatrove_jql_annotator.py [hunk 6]
```diff
@@ -0,0 +1,714 @@
...
+            if self.device_overwrite is None:
+                device_count = cuda.device_count()
+                cuda_device_id = rank % device_count
+                device = f'cuda:{cuda_device_id}'
+            else:
+                device = f'cuda:{self.device_overwrite}'
+
+        self.regression_heads = {}
+        for name, path in self.regression_head_checkpoints.items():
+            self.regression_heads[name] = RegressionHead.load_from_checkpoint(path, map_location=device).to(bfloat16)
```
[the remark this answers is not in the document]
Comment (ajude2s) [reply src/ml_filter/annotation/datatrove_jql_annotator.py]: I agree. 
We will need one dtype for each of the following:
1. Model precision
2. Precision at which embeddings are saved.
3. Precision at which scores/labels are saved.

And parameterize all of them in the config. I would say, always save embeddings in float32. Model precision and score/label precision can be parameterized? What do you think? @AbasKhan @fromm-m

File: src/ml_filter/annotation/regression_head.py (hunk 5 above)
Comment (ajude2s) [reply src/ml_filter/annotation/regression_head.py]: This is original JQL regression head code from HessanAI:
https://github.com/JQL-AI/JQL-Annotation-Pipeline/tree/56e5f3a4d687a06894c17fa57fa59b032e97a273/src/utils

File: src/ml_filter/annotation/datatrove_jql_annotator.py [hunk 7]
```diff
@@ -0,0 +1,714 @@
...
+    Returns:
+        dict: A metadata dictionary safe for JSON serialization, without the document text.
+    """
+
+    def safe_json(val):
+        if isinstance(val, bytes):
+            return val.decode("utf-8", errors="ignore")
+        return val
+
+    data = {key: safe_json(val) for key, val in dataclasses.asdict(document).items() if val and key != "text"}
```
[the remark this answers is not in the document]
Comment (ajude2s) [reply src/ml_filter/annotation/datatrove_jql_annotator.py]: No, the `text` here does not correspond to the text field in jsonl file. This `text` here corresponds to a attribute of the datatrove dataclass `Document`

File: src/ml_filter/annotation/datatrove_jql_annotator.py [hunk 8]
```diff
@@ -0,0 +1,714 @@
...
+
+    def _write_batch(self, filename: str):
+        if not self._batches[filename]:
+            return
+
+        batch = self._batches.pop(filename)
+        embeddings = np.stack([doc["metadata"]["embedding"] for doc in batch], dtype=np.float32)
+        document_id = [doc["metadata"]["document_id"] for doc in batch]
+        if self.save_labels and "label" in batch[0]["metadata"]:
+            labels = np.array([doc["metadata"]["label"] for doc in batch], dtype=np.float32)
```
[the remark this answers is not in the document]
Comment (ajude2s) [reply src/ml_filter/annotation/datatrove_jql_annotator.py]: Fixed in 17ebd6370e3e8e8077384849876f40e68ab25248

File: src/ml_filter/annotation/datatrove_jql_annotator.py [hunk 9]
```diff
@@ -0,0 +1,714 @@
...
+        file = self._writers[filename]
+        group_name = self.dataset_name
+
+        # If group does not exist, create fresh datasets (no duplication)
+        if group_name not in file:
+            group = file.create_group(group_name)
+            maxshape_emb = (None, embeddings.shape[1])
+            maxshape_ids = (None,)
+            emb_ds = group.create_dataset(
+                "embeddings", shape=(0, embeddings.shape[1]), maxshape=maxshape_emb, compression="gzip", dtype=np.float32
```
[the remark this answers is not in the document]
Comment (ajude2s) [reply src/ml_filter/annotation/datatrove_jql_annotator.py]: Fixed in 17ebd6370e3e8e8077384849876f40e68ab25248

File: src/ml_filter/annotation/datatrove_jql_annotator.py [hunk 10]
```diff
@@ -0,0 +1,714 @@
...
+    """
+    A pipeline step that embeds batches of documents using a specified embedding model.
+
+    This step supports GPU acceleration and handles device assignment across multiple ranks
+    (e.g., in distributed settings). Each document receives an `embedding` field in metadata
+    containing the embedding vector.
+
+    Args:
+        embedder_model_id: HuggingFace model ID or local path to the embedding model.
+        Default is Snowflake Arctic embed v2.0.
```
[the remark this answers is not in the document]
Comment (ajude2s) [reply src/ml_filter/annotation/datatrove_jql_annotator.py]: Fixed in 17ebd6370e3e8e8077384849876f40e68ab25248

File: src/ml_filter/annotation/datatrove_jql_annotator.py [hunk 11]
```diff
@@ -0,0 +1,714 @@
...
+from torch import bfloat16, cuda, no_grad
+
+from ml_filter.annotation.embedder import get_embedder_instance
+from ml_filter.data_processing.hash_data_files import read_existing_hashes
+from ml_filter.annotation.regression_head import RegressionHead
+
+load_dotenv()
+
+
+def find_max_batch_size(embedder, max_limit=50000, step=100):
```
[the remark this answers is not in the document]
Comment (ajude2s) [reply src/ml_filter/annotation/datatrove_jql_annotator.py]: Removed obsolete function in 62f3fe79a8e91b25e3fcaeef024684a1557c9bac

File: src/ml_filter/annotation/datatrove_jql_annotator.py [hunk 12]
```diff
@@ -0,0 +1,714 @@
...
+
+    name = "🐿 Jsonl"
+    _requires_dependencies = ["orjson"]
+
+    def __init__(
+        self,
+        data_folder: DataFolderLike,
+        keys_to_index: list[str],
+        paths_file: DataFileLike | None = None,
+        compression: Literal["infer", "gzip", "zstd"] | None = "infer",
```
[the remark this answers is not in the document]
Comment (ajude2s) [reply src/ml_filter/annotation/datatrove_jql_annotator.py]: Fixed

File: src/ml_filter/annotation/embedder.py (hunk 4 above)
Comment (ajude2s) [reply src/ml_filter/annotation/embedder.py]: Um, this was merged from your branch related to the Qwen3Embedder. Is this function needed for inference? @AbasKhan

File: src/ml_filter/annotation/embedder.py [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+
+        with torch.no_grad(), torch.cuda.device(self.device):  # Disable gradient calculation for inference
+            output = self.model(**batch_tokens)
+            # Extract embeddings using last token pooling strategy
+            embeddings = self.last_token_pool(output.last_hidden_state, batch_tokens["attention_mask"])
+
+            # Apply L2 normalization to the embeddings
+            embeddings = F.normalize(embeddings, p=2, dim=1)
+
+        return embeddings.cpu().tolist()
```
[the remark this answers is not in the document]
Comment (ajude2s) [reply src/ml_filter/annotation/embedder.py]: This is fixed now. There is only one `.cpu().tolist()` happening.

File: src/ml_filter/annotation/embedder.py (hunk 4 above)
Comment (AbasKhan) [reply src/ml_filter/annotation/embedder.py]: The function isnt used anywhere anymore, i think its an artifact from qwens inference code for getting semantic similarity between given strings. We can safely remove it

File: src/ml_filter/annotation/regression_head.py (hunk 5 above)
Comment (AbasKhan) [reply src/ml_filter/annotation/regression_head.py]: I think the code at `https://github.com/Modalities/ml_filter/blob/regression_exp_abbas_embed/src/ml_filter/models/annotator_model_head.py` does exactly the same thing now right ?, so we dont need both of them ?. On the first look they look the same to me

**Review by AbasKhan — CHANGES_REQUESTED** [review]

File: src/ml_filter/annotation/embedder.py (hunk 4 above)
Comment (ajude2s) [reply src/ml_filter/annotation/embedder.py]: Removed in d6ec90caf280764b959576c6b5d74cfa8eb8c0ae

File: src/ml_filter/annotation/regression_head.py (hunk 5 above)
Comment (ajude2s) [reply src/ml_filter/annotation/regression_head.py]: This would be potential blocker for this PR and has direct dependency to #233. 
We have used HessanAI checkpoints so far for annotation and they are in PyTorch Lightning format. The new regression heads trained in #233 cannot be loaded here unless it is converted to HF format. The PR #233 should be merged first in this case. @AbasKhan @fromm-m
